### PR TITLE
Creatures above cast shadows onto tiles below

### LINKED ISSUE #66657: The ability to see flying creatures as well as those under the ground
### Is your feature request related to a problem? Please describe.

When you walk on the surface, the flying creatures are completely invisible. If they are wasps, for example, the compass is absolutely useless. So you have to press x and then press all levels up to see who flies. But the problem is that if you do not know about it or there is no hive - a flying creature is absolutely invisible, for example in a crowd of monsters. It is not on the minicard and it does not appear on the screen. 

### Solution you would like.

It would be good to make the monsters displayed, but with a level of darkening corresponding to their position on Z level, and if they are higher - that the character and the monster could pass through each other. And then I think the problem will be solved

### Describe alternatives you have considered.

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rewryte**
#### Summary
Features "Creatures above cast shadows onto tiles below"

#### Purpose of change
Creatures above the z-level of the player can be difficult to detect due to the lack of visual indications of their presence.
Fixes #66657 

#### Describe the solution
Make creatures cast shadows unto tiles below.  Image used for the shadow is determined by each tileset via the new `shadow` sprite id.  The hostility indicator will also be drawn when applicable.  Naturally, this feature will not be functional until a `shadow` sprite is assigned to the tilesets and the next tileset update takes place.  I will work on that unless someone else wants to.

shadowtest

A simpler version is used in curses which mimics how critters below the current z-level are currently drawn.  Because the "critter above" indicator has the potential to obscure the underlying tile, both the "critter above" and "critter below" indicator were made to blink.  These are affected by the "Blinking effects speed" setting.

curses

#### Testing
Builds ok in both tiles and curses.
Tested in both isometric and non-isometric mode by assigning a `shadow` sprite using a mod
Both shadow and hostility indicators are drawn correctly.
Performance impact is negligible.

### CODE REVIEW

**Comment by GuardianDll** [comment]
I guess it should resolve #66657

**Comment by Rewryte** [comment]
> I guess it should resolve #66657

Thanks.  Updated main comment.

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -1762,6 +1768,13 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
                                     // If no vpart drawn, revert height_3d changes
                                     p.com.height_3d = temp_height_3d;
                                 }
+                            } else if( f == &cata_tiles::draw_critter_at ) {
+                                // Draw
+                                if( !( this->*f )( draw_loc, var->ll, p.com.height_3d, var->invisible, false ) && do_draw_shadow &&
+                                    cur_zlevel == p.draw_min_z ) {
+                                    // Draw shadow of flying critters on bottom-most tile if no other critter drawn
+                                    draw_critter_above( draw_loc, p.ll, p.height_3d, p.invisible );
```
[the remark this answers is not in the document]
Comment (Rewryte) [reply src/cata_tiles.cpp:1776]: Thanks.  I already fixed these on my end.  Just waiting for clang tidy to finish before I push.

**Comment by RenechCDDA** [comment]
I hope you don't mind some questions

How does it handle if a creature(with its own sprite) is occupying the tile where the shadow would be placed? (Looking at the changes, seems it just doesn't draw the shadow?)

What if the creature on z=0 does not have a sprite defined (or is using ASCII fallback), does it still cancel out drawing the shadow sprite?

Does it draw the shadow if the shadow would be placed on an 'impassable' tile (e.g. the walled edge of a building)?

**Comment by Rewryte** [comment]
> How does it handle if a creature(with its own sprite) is occupying the tile where the shadow would be placed? (Looking at the changes, seems it just doesn't draw the shadow?)

Yup.  Tilesets like Ultica give creatures their own shadows so I didn't think it necessary.

> What if the creature on z=0 does not have a sprite defined (or is using ASCII fallback), does it still cancel out drawing the shadow sprite?

Any draw function with a fallback should still return true which should stop the shadow drawing.  If nothing is drawn, such as in the case of an invisible creature, then the shadow function will run.

> Does it draw the shadow if the shadow would be placed on an 'impassable' tile (e.g. the walled edge of a building)?

Nope.  Both walls and the ceiling above them should return true on `dont_draw_lower_floor`.  This stops the function from scanning for creatures.

**Pull request merged**